# Recommendation Engine

In practice, recommendation engines are of three kinds:
 1. **Popularity-based** engines: Usually the most simple to implement be also the most impersonal
 2. **Content-based** engines: The recommendations are based on the description of the products
 3. **Collaborative filtering** engines: Records from various users provide recommendations based on user similarities
 
 
 
1. Basic functioning of the engine 
    * 1.1 Similarity
    * 1.2 Popularity
2. Definition of the recommendation engine functions
3. Making meaningfull recommendations


## A Simple Recommener Engine

The Simple Recommender offers generalized recommnendations to every user based on movie popularity and (sometimes) genre. The basic idea behind this recommender is that movies that are more popular and more critically acclaimed will have a higher probability of being liked by the average audience. This model does not give personalized recommendations based on the user. 

The implementation of this model is extremely trivial. All we have to do is sort our movies based on ratings and popularity and display the top movies of our list. As an added step, we can pass in a genre argument to get the top movies of a particular genre. 

In [1]:
from ast import literal_eval
import pandas as pd
md = pd. read_csv('movies_metadata.csv')

md['genres'] = md['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
md.head(2)

/Users/pymacbit/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3051: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                         genres                              homepage    id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   862   
1  [Adventure, Fantasy, Family]                                   NaN  8844   

     imdb_id original_language original_title  \
0  tt0114709                en      Toy Story   
1  tt0113497                en        Jumanji   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

     status                                    tagline      title  video  \
0  Released                                        NaN  Toy Story  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   

  vote_average vote_count  
0          7.7     5415.0  
1          6.9     2413.0  

[2 rows x 24 columns]

I use the TMDB Ratings to come up with our **Top Movies Chart.** I will use IMDB's *weighted rating* formula to construct my chart. Mathematically, it is represented as follows:

Weighted Rating (WR) = $(\frac{v}{v + m} . R) + (\frac{m}{v + m} . C)$

where,
* *v* is the number of votes for the movie
* *m* is the minimum votes required to be listed in the chart
* *R* is the average rating of the movie
* *C* is the mean vote across the whole report

The next step is to determine an appropriate value for *m*, the minimum votes required to be listed in the chart. We will use **95th percentile** as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.

I will build our overall Top 250 Chart and will define a function to build charts for a particular genre. Let's begin!

In [2]:
vote_counts = md[md['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = md[md['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
print ("Vote Count Average Mean:   "  , C)
m = vote_counts.quantile(0.95)
print ("Vote Count Quantile 0.95:  ", m)

Vote Count Average Mean:    5.244896612406511
Vote Count Quantile 0.95:   434.0


In [3]:
import numpy as np

md['year'] = pd.to_datetime(md['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

qualified = md[(md['vote_count'] >= m) & (md['vote_count'].notnull()) & (md['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified.describe()

vote_count  vote_average
count   2274.000000   2274.000000
mean    1582.002199      6.159631
std     1570.502308      0.839273
min      434.000000      2.000000
25%      628.000000      6.000000
50%      984.000000      6.000000
75%     1831.000000      7.000000
max    14075.000000      9.000000

Therefore, to qualify to be considered for the chart, a movie has to have at least **434 votes** on TMDB. We also see that the average rating for a movie on TMDB is **5.244** on a scale of 10. **2274** 
And, Movies qualify to be on our chart.


In [4]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

qualified['wr'] = qualified.apply(weighted_rating, axis=1)

In [5]:
qualified = qualified.sort_values('wr', ascending=False).head(250)
qualified.head()

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
22879                                       Interstellar  2014       11187   
2843                                          Fight Club  1999        9678   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   

       vote_average popularity  \
15480             8    29.1081   
12481             8    123.167   
22879             8    32.2135   
2843              8    63.8696   
4863              8    32.0707   

                                                  genres        wr  
15480  [Action, Thriller, Science Fiction, Mystery, A...  7.917588  
12481                   [Drama, Action, Crime, Thriller]  7.905871  
22879                [Adventure, Drama, Science Fiction]  7.897107  
2843                                             [Drama]  7.881753  
4863                        [Adventure, Fantasy, Action]  7.871787

In [6]:
s = md.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'genre'
gen_md = md.drop('genres', axis=1).join(s)

In [7]:
def build_chart(genre, percentile=0.85):
    df = gen_md[gen_md['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    return qualified

## Top Action Movies

In [8]:
build_chart('Action').head(10)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
7000       The Lord of the Rings: The Return of the King  2003        8226   
5814               The Lord of the Rings: The Two Towers  2002        7641   
256                                            Star Wars  1977        6778   
1154                             The Empire Strikes Back  1980        5998   
4135                                            Scarface  1983        3017   
9430                                              Oldboy  2003        2000   
1910                                       Seven Samurai  1954         892   

       vote_average popularity        wr  
15480             8    29.1081  7.955099  
12481             8    123.167  7.948610  
4863              8    32.0707  7.929579  
7000              8    29.3244  7.924031  
5814              8    29.4235  7.918382  
256               8    42.1497  7.908327  
1154              8     19.471  7.896841  
4135              8    11.2997  7.802046  
9430              8    10.6169  7.711649  
1910              8    15.0178  7.426145

## Top Sci-Fi Movies

In [9]:
build_chart('Science Fiction').head(10)

title  year  vote_count  vote_average popularity  \
15480                Inception  2010       14075             8    29.1081   
22879             Interstellar  2014       11187             8    32.2135   
256                  Star Wars  1977        6778             8    42.1497   
1225        Back to the Future  1985        6239             8    25.7785   
1154   The Empire Strikes Back  1980        5998             8     19.471   
1163        A Clockwork Orange  1971        3432             8    17.1126   
1901                Metropolis  1927         666             8    14.4879   
14551                   Avatar  2009       12114             7    185.071   
17818             The Avengers  2012       12000             7    89.8876   
23753  Guardians of the Galaxy  2014       10014             7    53.2916   

             wr  
15480  7.939069  
22879  7.923728  
256    7.876106  
1225   7.865868  
1154   7.860722  
1163   7.764533  
1901   7.078592  
14551  6.952299  
17818  6.951856  
23753  6.942571

## Top Animated Movies

In [10]:
build_chart('Animation').head(15)

title  year  vote_count  vote_average  \
359                      The Lion King  1994        5520             8   
5481                     Spirited Away  2001        3968             8   
9698              Howl's Moving Castle  2004        2049             8   
2884                 Princess Mononoke  1997        2041             8   
5833                My Neighbor Totoro  1988        1730             8   
40251                       Your Name.  2016        1030             8   
5553            Grave of the Fireflies  1988         974             8   
19901                         Paperman  2012         734             8   
39386                            Piper  2016         487             8   
20779                    Wolf Children  2012         483             8   
25044                  Song of the Sea  2014         420             8   
31658                            Feast  2014         420             8   
19101                           Presto  2008         371             8   
24209  The Tale of the Princess Kaguya  2013         350             8   
13724                               Up  2009        7048             7   

      popularity        wr  
359      21.6058  7.909339  
5481     41.0489  7.875933  
9698      16.136  7.772103  
2884     17.1667  7.771305  
5833     13.5073  7.735274  
40251  34.461252  7.589820  
5553    0.010902  7.570962  
19901    7.19863  7.465676  
39386  11.243161  7.285132  
20779    10.2495  7.281198  
25044    6.96736  7.212999  
31658    7.36566  7.212999  
19101    8.56386  7.150295  
24209    7.56605  7.120255  
13724    19.3309  6.961798

## Content Based Recommender

The recommender we built in the previous section suffers some severe limitations. For one, it gives the same recommendation to everyone, regardless of the user's personal taste. If a person who loves romantic movies (and hates action) were to look at our Top 15 Chart, s/he wouldn't probably like most of the movies. If s/he were to go one step further and look at our charts by genre, s/he wouldn't still be getting the best recommendations.

For instance, consider a person who loves *Dilwale Dulhania Le Jayenge*, *My Name is Khan* and *Kabhi Khushi Kabhi Gham*. One inference we can obtain is that the person loves the actor Shahrukh Khan and the director Karan Johar. Even if s/he were to access the romance chart, s/he wouldn't find these as the top recommendations.

To personalise our recommendations more, I am going to build an engine that computes similarity between movies based on certain metrics and suggests movies that are most similar to a particular movie that a user liked. Since we will be using movie metadata (or content) to build this engine, this also known as **Content Based Filtering.**

I will build two Content Based Recommenders based on:
* Movie Overviews and Taglines
* Movie Cast, Crew, Keywords and Genre

Also, as mentioned in the introduction, I will be using a subset of all the movies available to us due to limiting computing power available to me. 

In [11]:
links_small = pd.read_csv('links_small.csv')
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')
md = md.drop([19730, 29503, 35587])
md['id'] = md['id'].astype('int')

In [12]:
smd = md[md['id'].isin(links_small)]
smd.shape

(9099, 25)

In [13]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

/Users/pymacbit/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/Users/pymacbit/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/Users/pymacbit/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = va

In [14]:
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
from surprise import Reader, Dataset, SVD, evaluate
import warnings; warnings.simplefilter('ignore')

tf = TfidfVectorizer(analyzer='word', ngram_range=(1, 2),min_df=0, stop_words='english')
tfidf_matrix = tf.fit_transform(smd['description'])

#### Cosine Similarity

I will be using the Cosine Similarity to calculate a numeric quantity that denotes the similarity between two movies. Mathematically, it is defined as follows:

$cosine(x,y) = \frac{x. y^\intercal}{||x||.||y||} $

Since we have used the TF-IDF Vectorizer, calculating the Dot Product will directly give us the Cosine Similarity Score. Therefore, we will use sklearn's **linear_kernel** instead of cosine_similarities since it is much faster.

In [15]:
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

 We now have a pairwise cosine similarity matrix for all the movies in our dataset. The next step is to write a function that returns the 30 most similar movies based on the cosine similarity score.

In [16]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return pd.DataFrame(titles.iloc[movie_indices])

In [17]:
get_recommendations('The Godfather').head(10)

title
973    The Godfather: Part II
8387               The Family
3509                     Made
4196       Johnny Dangerously
29             Shanghai Triad
5667                     Fury
2412           American Movie
1582  The Godfather: Part III
4221                  8 Women
2159            Summer of Sam

In [18]:
get_recommendations('The Dark Knight').head(10)

title
7931                    The Dark Knight Rises
132                            Batman Forever
1113                           Batman Returns
8227  Batman: The Dark Knight Returns, Part 2
7565               Batman: Under the Red Hood
524                                    Batman
7901                         Batman: Year One
2579             Batman: Mask of the Phantasm
2696                                      JFK
8165  Batman: The Dark Knight Returns, Part 1

Now, thats a pattern. We could observe the **Batman** Movies all in one place here. And as matter of fact I love the **Batmat Triology <3**

### Metadata Based Recommender

To build our standard metadata based content recommender, we will need to merge our current dataset with the crew and the keyword datasets. Let us prepare this data as our first step.

In [19]:
credits = pd.read_csv('credits.csv')
keywords = pd.read_csv('keywords.csv')
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
md['id'] = md['id'].astype('int')
md = md.merge(credits, on='id')
md = md.merge(keywords, on='id')

We now have our cast, crew, genres and credits, all in one dataframe. Let us wrangle this a little more using the following intuitions:

1. **Crew:** From the crew, we will only pick the director as our feature since the others don't contribute that much to the *feel* of the movie.
2. **Cast:** Choosing Cast is a little more tricky. Lesser known actors and minor roles do not really affect people's opinion of a movie. Therefore, we must only select the major characters and their respective actors. Arbitrarily we will choose the top 3 actors that appear in the credits list. 

In [20]:
smd = md[md['id'].isin(links_small)]
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x))
smd['crew_size'] = smd['crew'].apply(lambda x: len(x))

In [21]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [22]:
smd['director'] = smd['crew'].apply(get_director)
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
smd['cast'] = smd['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

What I plan on doing is creating a metadata dump for every movie which consists of **genres, director, main actors and keywords.** I then use a **Count Vectorizer** to create our count matrix as we did in the Description Recommender. The remaining steps are similar to what we did earlier: we calculate the cosine similarities and return movies that are most similar.

These are steps I follow in the preparation of my genres and credits data:
1. **Strip Spaces and Convert to Lowercase** from all our features. This way, our engine will not confuse between **Johnny Depp** and **Johnny Galecki.** 
2. **Mention Director 3 times** to give it more weight relative to the entire cast.

In [23]:
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [24]:
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
smd['director'] = smd['director'].apply(lambda x: [x,x, x])

## Keywords

We will do a small amount of pre-processing of our keywords before putting them to any use. As a first step, we calculate the frequenct counts of every keyword that appears in the dataset.

In [25]:
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [26]:
s = s.value_counts()
s[:5]

independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: keyword, dtype: int64

Keywords occur in frequencies ranging from 1 to 610. We do not have any use for keywords that occur only once. Therefore, these can be safely removed. Finally, we will convert every word to its stem so that words such as *Dogs* and *Dog* are considered the same.

In [27]:
s = s[s > 1]

In [28]:
stemmer = SnowballStemmer('english')
stemmer.stem('dogs')

'dog'

In [29]:
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words

In [30]:
smd['keywords'] = smd['keywords'].apply(filter_keywords)
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [31]:
smd['soup'] = smd['keywords'] + smd['cast'] + smd['director'] + smd['genres']
smd['soup'] = smd['soup'].apply(lambda x: ' '.join(x))

count = CountVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

cosine_sim = cosine_similarity(count_matrix, count_matrix)

smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

We will reuse the get_recommendations function that we had written earlier. Since our cosine similarity scores have changed, we expect it to give us different (and probably better) results. Let us check for **Iron Man** again and see what recommendations I get this time around.

In [32]:
get_recommendations('Iron Man').head(10)

title
7600                  Iron Man 2
3536                        Made
8392                  Iron Man 3
6359  Zathura: A Space Adventure
7793            Cowboys & Aliens
8661                        Chef
4930                         Elf
7969                The Avengers
9028             The Jungle Book
8869                     Ant-Man

## Popularity and Ratings

One thing that we notice about our recommendation system is that it recommends movies regardless of ratings and popularity. It is true that **Batman and Robin** has a lot of similar characters as compared to **The Dark Knight** but it was a terrible movie that shouldn't be recommended to anyone.

Therefore, we will add a mechanism to remove bad movies and return movies which are popular and have had a good critical response.

I will take the top 25 movies based on similarity scores and calculate the vote of the 60th percentile movie. Then, using this as the value of $m$, we will calculate the weighted rating of each movie using IMDB's formula like we did in the Simple Recommender section.

In [33]:
def improved_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]
    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year']]
    vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(0.60)
    qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    qualified['wr'] = qualified.apply(weighted_rating, axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(10)
    return qualified

In [34]:
improved_recommendations('Iron Man')

title  vote_count  vote_average  year  \
7969                         The Avengers       12000             7  2012   
8712              Guardians of the Galaxy       10014             7  2014   
8872           Captain America: Civil War        7462             7  2016   
8868              Avengers: Age of Ultron        6908             7  2015   
8869                              Ant-Man        6029             7  2015   
8626  Captain America: The Winter Soldier        5881             7  2014   
7889                   X-Men: First Class        5252             7  2011   
8392                           Iron Man 3        8951             6  2013   
7923   Captain America: The First Avenger        7174             6  2011   
7600                           Iron Man 2        6969             6  2010   

            wr  
7969  6.938739  
8712  6.927095  
8872  6.903532  
8868  6.896252  
8869  6.882142  
8626  6.879380  
7889  6.866037  
8392  5.965081  
7923  5.956925  
7600  5.955732

In [35]:
improved_recommendations("Hulk")

title  vote_count  vote_average  year        wr
7969              The Avengers       12000             7  2012  6.938739
7009                  Iron Man        8951             7  2008  6.918837
8869                   Ant-Man        6029             7  2015  6.882142
8258                Life of Pi        5912             7  2012  6.879969
7889        X-Men: First Class        5252             7  2011  6.866037
7600                Iron Man 2        6969             6  2010  5.955732
3049                     X-Men        4172             6  2000  5.928850
8450             The Wolverine        4110             6  2013  5.927880
7286  X-Men Origins: Wolverine        4086             6  2009  5.927497
4635                        X2        3572             6  2003  5.918194

## Collaborative Filtering

Our content based engine suffers from some severe limitations. It is only capable of suggesting movies which are *close* to a certain movie. That is, it is not capable of capturing tastes and providing recommendations across genres.

Also, the engine that we built is not really personal in that it doesn't capture the personal tastes and biases of a user. Anyone querying our engine for recommendations based on a movie will receive the same recommendations for that movie, regardless of who s/he is.

Therefore, in this section, we will use a technique called **Collaborative Filtering** to make recommendations to Movie Watchers. Collaborative Filtering is based on the idea that users similar to a me can be used to predict how much I will like a particular product or service those users have used/experienced but I have not.

I will not be implementing Collaborative Filtering from scratch. Instead, I will use the **Surprise** library that used extremely powerful algorithms like **Singular Value Decomposition (SVD)** to minimise RMSE (Root Mean Square Error) and give great recommendations.

In [36]:
reader = Reader()
ratings = pd.read_csv('ratings_small.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [37]:
data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)
data.split(n_folds=5)

In [38]:
svd = SVD()
evaluate(svd, data, measures=['RMSE', 'MAE'])

Evaluating RMSE, MAE of algorithm SVD.

------------
Fold 1
RMSE: 0.8941
MAE:  0.6870
------------
Fold 2
RMSE: 0.8951
MAE:  0.6897
------------
Fold 3
RMSE: 0.8951
MAE:  0.6913
------------
Fold 4
RMSE: 0.9025
MAE:  0.6928
------------
Fold 5
RMSE: 0.8901
MAE:  0.6851
------------
------------
Mean RMSE: 0.8954
Mean MAE : 0.6892
------------
------------


CaseInsensitiveDefaultDict(list,
                           {'rmse': [0.8940958687337609,
                             0.8951490055427839,
                             0.8951343853251502,
                             0.9025282478837663,
                             0.8900956600375902],
                            'mae': [0.6869727275507541,
                             0.6897352059701999,
                             0.6913431646706059,
                             0.6928490313194954,
                             0.6850915257955962]})

We get a mean **Root Mean Sqaure Error** of 0.8963 which is more than good enough for our case. Let us now train on our dataset and arrive at predictions.

In [39]:
trainset = data.build_full_trainset()
svd.train(trainset)

In [40]:
ratings[ratings['userId'] == 1]

userId  movieId  rating   timestamp
0        1       31     2.5  1260759144
1        1     1029     3.0  1260759179
2        1     1061     3.0  1260759182
3        1     1129     2.0  1260759185
4        1     1172     4.0  1260759205
5        1     1263     2.0  1260759151
6        1     1287     2.0  1260759187
7        1     1293     2.0  1260759148
8        1     1339     3.5  1260759125
9        1     1343     2.0  1260759131
10       1     1371     2.5  1260759135
11       1     1405     1.0  1260759203
12       1     1953     4.0  1260759191
13       1     2105     4.0  1260759139
14       1     2150     3.0  1260759194
15       1     2193     2.0  1260759198
16       1     2294     2.0  1260759108
17       1     2455     2.5  1260759113
18       1     2968     1.0  1260759200
19       1     3671     3.0  1260759117

In [41]:
svd.predict(1, 302, 3)

Prediction(uid=1, iid=302, r_ui=3, est=2.8338843515285683, details={'was_impossible': False})

For movie with ID 302, we get an estimated prediction of **2.686**. One startling feature of this recommender system is that it doesn't care what the movie is (or what it contains). It works purely on the basis of an assigned movie ID and tries to predict ratings based on how the other users have predicted the movie.

## Hybrid Recommender

Another approach to combining collaborative and content-based filtering is to make predictions based on a weighted average of the content-based recommendation and the collaborative recommendation.  The rank of each item being recommended could be a measure for the weight. In this way the highest recommendation receives the highest 


In [42]:
improved_recommendations('Shutter Island')

title  vote_count  vote_average  year        wr
101               Taxi Driver        2632             8  1976  7.610008
8544  The Wolf of Wall Street        6768             7  2013  6.894236
6599             The Departed        4455             7  2006  6.844198
8007                     Hugo        2197             7  2011  6.710485
4439        Gangs of New York        1964             7  2002  6.682354
6047              The Aviator        1526             7  2004  6.611370
15                     Casino        1343             7  1995  6.571348
1001              Raging Bull         968             7  1980  6.456694
1104                Cape Fear         692             7  1991  6.323521
5376             Mean Streets         359             7  1973  6.039452

In this section, I will try to build a simple hybrid recommender that brings together techniques we have implemented in the content based and collaborative filter based engines. This is how it will work:

* **Input:** User ID and the Title of a Movie
* **Output:** Similar movies sorted on the basis of expected ratings by that particular user.

In [43]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [44]:
id_map = pd.read_csv('links_small.csv')[['movieId', 'tmdbId']]
id_map['tmdbId'] = id_map['tmdbId'].apply(convert_int)
id_map.columns = ['movieId', 'id']
id_map = id_map.merge(smd[['title', 'id']], on='id').set_index('title')
#id_map = id_map.set_index('tmdbId')

In [45]:
indices_map = id_map.set_index('id')

In [46]:
def hybrid(userId, title):
    idx = indices[title]
    tmdbId = id_map.loc[title]['id']
    #print(idx)
    movie_id = id_map.loc[title]['movieId']
    
    sim_scores = list(enumerate(cosine_sim[int(idx)]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year', 'id']]
    movies['est'] = movies['id'].apply(lambda x: svd.predict(userId, indices_map.loc[x]['movieId']).est)
    movies = movies.sort_values('est', ascending=False)
    return movies.head(10)

In [47]:
hybrid(1, 'Iron Man')

title  vote_count  vote_average  year  \
8626  Captain America: The Winter Soldier      5881.0           7.6  2014   
8712              Guardians of the Galaxy     10014.0           7.9  2014   
7969                         The Avengers     12000.0           7.4  2012   
8868              Avengers: Age of Ultron      6908.0           7.3  2015   
7889                   X-Men: First Class      5252.0           7.1  2011   
3049                                X-Men      4172.0           6.8  2000   
4635                                   X2      3572.0           6.8  2003   
8869                              Ant-Man      6029.0           7.0  2015   
6493                X-Men: The Last Stand      3581.0           6.3  2006   
7923   Captain America: The First Avenger      7174.0           6.6  2011   

          id       est  
8626  100402  3.079231  
8712  118340  3.049003  
7969   24428  3.014135  
8868   99861  3.002582  
7889   49538  2.969639  
3049   36657  2.916742  
4635   36658  2.834102  
8869  102899  2.771176  
6493   36668  2.769405  
7923    1771  2.690685

We see that for our hybrid recommender, we get different recommendations for different users although the movie is the same. Hence, our recommendations are more personalized and tailored towards particular users.

# Conclusion

In this notebook, I have built 4 different recommendation engines based on different ideas and algorithms. They are as follows:

1. **Simple Recommender:** This system used overall TMDB Vote Count and Vote Averages to build Top Movies Charts, in general and for a specific genre. The IMDB Weighted Rating System was used to calculate ratings on which the sorting was finally performed.
2. **Content Based Recommender:** We built two content based engines; one that took movie overview and taglines as input and the other which took metadata such as cast, crew, genre and keywords to come up with predictions. We also deviced a simple filter to give greater preference to movies with more votes and higher ratings.
3. **Collaborative Filtering:** We used the powerful Surprise Library to build a collaborative filter based on single value decomposition. The RMSE obtained was less than 1 and the engine gave estimated ratings for a given user and movie.
4. **Hybrid Engine:** We brought together ideas from content and collaborative filterting to build an engine that gave movie suggestions to a particular user based on the estimated ratings that it had internally calculated for that user


    Acknowledgement: 
    - The Movie Dataset Kaggle
    - Dataset contributors Fabien Daniel 
    - Rounak Banik for guidance and the tutorials